CNDB-9193 Track sensor bytes for LWT

aymkhalil opened the pull request:
Addresses: https://github.com/riptano/cndb/issues/9193

Bare minimum PR to track sensor byes for LWT. Notes:
* `WRITE_BYTES` & `READ_BYTES` are utilized for tracking LWT (both the Paxos system table and the effective read/write on the user table). 
* Read/write bytes are transferred from the Paxos system table to the user table that initiated Paxos. This way we avoid special logic in handling sensors via `TrackingRowIterator` or `Mutation#apply` because normally, the sensors in those table are tied with the system table context (not the user table)
* Although the code path for the effective Paxos commit on the user table, as long as we have `RequestSensors` registered in the Paxos verb handler, counting bytes is business as usual (share code with regular writes). The final sensor value will be the sum of what's written in the user table + the Paxos system table. 
* Similarly for Paxos reads (i.e. with `CL=SERIAL`), the `ReadCommandVerbHandler` will be invoked if Paxos succeeds and the read bytes will be updated accordingly. The final sensor value will be the sum of bytes read from the user table + what's read from the Paxos system table.

aymkhalil commented:
Merged branch `cc-ucu` into `cc-ucu-lwt` and adjusted the PR to target `cc-ucu` https://github.com/datastax/cassandra/pull/1074/commits/dc53ad38ecc1cb3f3cfb36a68e3b036ba20a53a4

maxtomassi commented on test/unit/org/apache/cassandra/net/SensorsCustomParamsTest.java:
```diff
@@ -95,4 +118,72 @@ public void testIndexReadBytesTableParam()
     }
+
+    @Test
+    public void testAddWriteSensorToResponse()
```
Since the two tests are pretty much identical I was wondering if we can extract most of the logic from them and do something similar to the following:

```dtd
    public void testAddWriteSensorToResponse()
    {
        testAddSensorToResponse(Type.WRITE_BYTES,
               c -> SensorsCustomParams.encodeTableInWriteBytesRequestParam(c.getTable()),
               c -> SensorsCustomParams.encodeTableInWriteBytesTableParam(c.getTable()));
    }

    @Test
    public void testAddReadSensorToResponse()
    {
        testAddSensorToResponse(Type.READ_BYTES,
                ignored -> SensorsCustomParams.READ_BYTES_REQUEST,
                ignored -> SensorsCustomParams.READ_BYTES_TABLE);
    }

    private void testAddSensorToResponse(Type sensorType, Function<Context, String> requestParamFunc, Function<Context, String> tableParamContext)
    {
        RequestSensors sensors = new RequestSensors();
        UUID tableId = UUID.randomUUID();
        KeyspaceMetadata ksm = KeyspaceMetadata.create("ks1", null);
        TableMetadata tm = TableMetadata.builder("ks1", "t1", TableId.fromString(tableId.toString()))
                                        .addPartitionKeyColumn("pk", AsciiType.instance)
                                        .build();
        SensorsRegistry.instance.onCreateKeyspace(ksm);
        SensorsRegistry.instance.onCreateTable(tm);

        Context context = new Context("ks1", "t1", tableId.toString());
        sensors.registerSensor(context, sensorType);
        sensors.incrementSensor(context, sensorType, 13.0);
        sensors.syncAllSensors();

        Message.Builder<NoPayload> builder =
        Message.builder(Verb._TEST_1, noPayload)
               .withId(1);

        switch (sensorType)
        {
            case READ_BYTES:
                SensorsCustomParams.addReadSensorToResponse(builder, sensors, context);
                break;
            case WRITE_BYTES:
                SensorsCustomParams.addWriteSensorToResponse(builder, sensors, context);
                break;
            default:
                throw new IllegalArgumentException("Unexpected sensor type " + sensorType);
        }

        Message<NoPayload> msg = builder.build();
        assertNotNull(msg.header.customParams());
        assertEquals(2, msg.header.customParams().size());
        String requestParam = requestParamFunc.apply(context);
        String tableParam = tableParamContext.apply(context);
        assertTrue(msg.header.customParams().containsKey(requestParam));
        assertTrue(msg.header.customParams().containsKey(tableParam));
        double epsilon = 0.000001;
        assertEquals(13.0, SensorsCustomParams.sensorValueFromBytes(msg.header.customParams().get(requestParam)), epsilon);
        assertEquals(13.0, SensorsCustomParams.sensorValueFromBytes(msg.header.customParams().get(tableParam)), epsilon);
    }
```

aymkhalil replied:
I usually the "Rule of Three" for code duplication but really no strong opinions here.

maxtomassi commented on test/unit/org/apache/cassandra/sensors/SensorsWriteTest.java:
```diff
@@ -274,6 +287,92 @@ public void testSingleCounterMutation() throws WriteTimeoutException
+
+        // handle the commit again, this time paxos has state because if the first proposal and read bytes will be populated
+        handlePaxosPrepare(proposal);
+        Sensor readSensor = SensorsTestUtil.getThreadLocalRequestSensor(context, Type.READ_BYTES);
```
Shall we then check that we had no `READ_BYTES` sensor (or with value zero) for the previous request?

maxtomassi commented on test/unit/org/apache/cassandra/sensors/SensorsWriteTest.java:
```diff
@@ -274,6 +287,92 @@ public void testSingleCounterMutation() throws WriteTimeoutException
+        assertThat(registryWriteSensor).isEqualTo(writeSensor);
+        assertResponseSensors(writeSensor.getValue(), registryWriteSensor.getValue(), CF_STANDARD);
+
+        // handle the commit again, this time paxos has state because if the first proposal and read bytes will be populated
```
I guess you meant `because of the first proposal`?

maxtomassi commented on test/unit/org/apache/cassandra/sensors/SensorsWriteTest.java:
```diff
@@ -274,6 +287,92 @@ public void testSingleCounterMutation() throws WriteTimeoutException
+
+        // handle the commit again, this time paxos has state because if the first proposal and read bytes will be populated
+        handlePaxosPropose(proposal);
+        Sensor readSensor = SensorsTestUtil.getThreadLocalRequestSensor(context, Type.READ_BYTES);
```
Shall we then check that we had no `READ_BYTES` sensor (or with value zero) for the previous request?

aymkhalil replied:
Yes! Applies to the `testLWTPrepare` test too.

maxtomassi commented on test/unit/org/apache/cassandra/sensors/SensorsWriteTest.java:
```diff
@@ -274,6 +287,92 @@ public void testSingleCounterMutation() throws WriteTimeoutException
+        assertThat(registryWriteSensor).isEqualTo(writeSensor);
+        assertResponseSensors(writeSensor.getValue(), registryWriteSensor.getValue(), CF_STANDARD);
+
+        // handle the commit again, this time paxos has state because if the first proposal and read bytes will be populated
```
I guess you meant `because of the first proposal`?

maxtomassi commented on test/unit/org/apache/cassandra/net/SensorsCustomParamsTest.java:
```diff
@@ -95,4 +118,72 @@ public void testIndexReadBytesTableParam()
+        assertNotNull(msg.header.customParams());
+        assertEquals(2, msg.header.customParams().size());
+        assertTrue(msg.header.customParams().containsKey(SensorsCustomParams.READ_BYTES_REQUEST));
+        assertTrue(msg.header.customParams().containsKey(SensorsCustomParams.READ_BYTES_REQUEST));
```
I guess this one should be `READ_BYTES_TABLE`

aymkhalil replied:
Yes fixed indirectly by the refactoring you suggested

maxtomassi commented on test/unit/org/apache/cassandra/net/SensorsCustomParamsTest.java:
```diff
@@ -95,4 +118,72 @@ public void testIndexReadBytesTableParam()
+        assertTrue(msg.header.customParams().containsKey(SensorsCustomParams.READ_BYTES_REQUEST));
+        double epsilon = 0.000001;
+        assertEquals(13.0, SensorsCustomParams.sensorValueFromBytes(msg.header.customParams().get(SensorsCustomParams.READ_BYTES_REQUEST)), epsilon);
+        assertEquals(13.0, SensorsCustomParams.sensorValueFromBytes(msg.header.customParams().get(SensorsCustomParams.READ_BYTES_REQUEST)), epsilon);
```
I guess this one should be `READ_BYTES_TABLE`

aymkhalil replied:
ditto

maxtomassi reviewed (commented):
Looks pretty good, left a few minor comments and one suggestion

aymkhalil commented:
> Looks pretty good, left a few minor comments and one suggestion

Thanks I addressed your comments here: https://github.com/datastax/cassandra/pull/1074/commits/e998f04d1f511c15c9ca498ba584c6a8ac1571a2

I plan to rebase one more time after merging #1081 to make things simple.

aymkhalil commented:
> I plan to rebase one more time after merging https://github.com/datastax/cassandra/pull/1081 to make things simple.

@maxtomassi this is done: https://github.com/datastax/cassandra/pull/1074/commits/e328b1accd1cf9bed9b0c09d9258e00a658b3389. PTAL.

maxtomassi commented:
@aymkhalil I noticed there are some non-Java files in this PR (e.g. `.toDelete`, `hs_err_pid80370.log`, the `audit` folder), are they related to these changes? Do we need to upload them to the repo?

maxtomassi commented on tools/bin/analyzecompactionlog:
```diff
@@ -45,4 +45,4 @@ fi
 
-# vi:ai sw=4 ts=4 tw=0 et
\ No newline at end of file
+# vi:ai sw=4 ts=4 tw=0 et
```
Do we need this change?

aymkhalil replied:
oh no, removing!

aymkhalil commented:
> @aymkhalil I noticed there are some non-Java files in this PR (e.g. `.toDelete`, `hs_err_pid80370.log`, the `audit` folder), are they related to these changes? Do we need to upload them to the repo?

Removed unwanted files https://github.com/datastax/cassandra/pull/1074/commits/f2f397137508764352f42d1dd5219032d014e1b9.

maxtomassi reviewed (approved):
LGTM

aymkhalil commented:
Added this follow up task based on our offline discussion https://github.com/riptano/cndb/issues/9665
